### ISSUE TITLE
How to make Yolov8 Object detection model predict multiple labels per bounding box ?

### ISSUE THREAD

**Issue by IGlace** [ultralytics/ultralytics#22982]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Following the logic and current implementation of Yolov8 object detection in ultralytics here is the detection loss used : 

```
class v8DetectionLoss:
    """Criterion class for computing training losses for YOLOv8 object detection."""

    def __init__(self, model, tal_topk: int = 10):  # model must be de-paralleled
        """Initialize v8DetectionLoss with model parameters and task-aligned assignment settings."""
        device = next(model.parameters()).device  # get model device
        h = model.args  # hyperparameters

        m = model.model[-1]  # Detect() module
        self.bce = nn.BCEWithLogitsLoss(reduction="none")
        self.hyp = h
        self.stride = m.stride  # model strides
        self.nc = m.nc  # number of classes
        self.no = m.nc + m.reg_max * 4
        self.reg_max = m.reg_max
        self.device = device

        self.use_dfl = m.reg_max > 1

        self.assigner = TaskAlignedAssigner(topk=tal_topk, num_classes=self.nc, alpha=0.5, beta=6.0)
        self.bbox_loss = BboxLoss(m.reg_max).to(device)
        self.proj = torch.arange(m.reg_max, dtype=torch.float, device=device)

    def preprocess(self, targets: torch.Tensor, batch_size: int, scale_tensor: torch.Tensor) -> torch.Tensor:
        """Preprocess targets by converting to tensor format and scaling coordinates."""
        nl, ne = targets.shape
        if nl == 0:
            out = torch.zeros(batch_size, 0, ne - 1, device=self.device)
        else:
            i = targets[:, 0]  # image index
            _, counts = i.unique(return_counts=True)
            counts = counts.to(dtype=torch.int32)
            out = torch.zeros(batch_size, counts.max(), ne - 1, device=self.device)
            for j in range(batch_size):
                matches = i == j
                if n := matches.sum():
                    out[j, :n] = targets[matches, 1:]
            out[..., 1:5] = xywh2xyxy(out[..., 1:5].mul_(scale_tensor))
        return out

    def bbox_decode(self, anchor_points: torch.Tensor, pred_dist: torch.Tensor) -> torch.Tensor:
        """Decode predicted object bounding box coordinates from anchor points and distribution."""
        if self.use_dfl:
            b, a, c = pred_dist.shape  # batch, anchors, channels
            pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype))
            # pred_dist = pred_dist.view(b, a, c // 4, 4).transpose(2,3).softmax(3).matmul(self.proj.type(pred_dist.dtype))
            # pred_dist = (pred_dist.view(b, a, c // 4, 4).softmax(2) * self.proj.type(pred_dist.dtype).view(1, 1, -1, 1)).sum(2)
        return dist2bbox(pred_dist, anchor_points, xywh=False)

    def __call__(self, preds: Any, batch: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:
        """Calculate the sum of the loss for box, cls and dfl multiplied by batch size."""
        loss = torch.zeros(3, device=self.device)  # box, cls, dfl
        feats = preds[1] if isinstance(preds, tuple) else preds
        pred_distri, pred_scores = torch.cat([xi.view(feats[0].shape[0], self.no, -1) for xi in feats], 2).split(
            (self.reg_max * 4, self.nc), 1
        )

        pred_scores = pred_scores.permute(0, 2, 1).contiguous()
        pred_distri = pred_distri.permute(0, 2, 1).contiguous()

        dtype = pred_scores.dtype
        batch_size = pred_scores.shape[0]
        imgsz = torch.tensor(feats[0].shape[2:], device=self.device, dtype=dtype) * self.stride[0]  # image size (h,w)
        anchor_points, stride_tensor = make_anchors(feats, self.stride, 0.5)

        # Targets
        targets = torch.cat((batch["batch_idx"].view(-1, 1), batch["cls"].view(-1, 1), batch["bboxes"]), 1)
        targets = self.preprocess(targets.to(self.device), batch_size, scale_tensor=imgsz[[1, 0, 1, 0]])
        gt_labels, gt_bboxes = targets.split((1, 4), 2)  # cls, xyxy
        mask_gt = gt_bboxes.sum(2, keepdim=True).gt_(0.0)

        # Pboxes
        pred_bboxes = self.bbox_decode(anchor_points, pred_distri)  # xyxy, (b, h*w, 4)
        # dfl_conf = pred_distri.view(batch_size, -1, 4, self.reg_max).detach().softmax(-1)
        # dfl_conf = (dfl_conf.amax(-1).mean(-1) + dfl_conf.amax(-1).amin(-1)) / 2

        _, target_bboxes, target_scores, fg_mask, _ = self.assigner(
            # pred_scores.detach().sigmoid() * 0.8 + dfl_conf.unsqueeze(-1) * 0.2,
            pred_scores.detach().sigmoid(),
            (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
            anchor_points * stride_tensor,
            gt_labels,
            gt_bboxes,
            mask_gt,
        )

        target_scores_sum = max(target_scores.sum(), 1)

        # Cls loss
        # loss[1] = self.varifocal_loss(pred_scores, target_scores, target_labels) / target_scores_sum  # VFL way
        loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE

        # Bbox loss
        if fg_mask.sum():
            target_bboxes /= stride_tensor
            loss[0], loss[2] = self.bbox_loss(
                pred_distri, pred_bboxes, anchor_points, target_bboxes, target_scores, target_scores_sum, fg_mask
            )

        loss[0] *= self.hyp.box  # box gain
        loss[1] *= self.hyp.cls  # cls gain
        loss[2] *= self.hyp.dfl  # dfl gain

        return loss * batch_size, loss.detach()  # loss(box, cls, dfl)
```

According to the code here, it use sigmoid for score predictions, which mean it support multiple label predictions per bounding box, but the [TaskAlignedAssigner]( https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/tal.py) only assign one class per grid.

And i was also reading an article recently that explain the multiple label objection detection support was introduce to yolo architecture since Yolov3, you can find the article [Here](https://www.mdpi.com/2504-4990/5/4/83) .

Now i have 2 questions : 
- Why did ultralytics decide to abandon the logic of multiple label prediction per bounding box altough it still support it in loss function ?
- Is it sufficient to modify the TaskAlignedAssigner to use multiple label per bounding box, or is it more complicated than that ?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @IGlace, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for sharing the detailed context and the loss snippet—this is a great architecture/design question around label assignment vs. prediction heads. 🤔🧠 This is an automated response to help us triage quickly, and an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help the team understand your target behavior (multi-label per box) and how you’re attempting to achieve it, please also include:
-Which Ultralytics version (`pip show ultralytics`) and install method you’re using
-Your dataset label format and an example image + label file showing a single box with multiple labels (or how you intend to represent that)
-Your exact training command or Python snippet (from data config through `model.train(...)`)
-A short expected-vs-actual description of what you want at inference time (e.g., one box with multiple class confidences retained vs. multiple boxes) 📦
-Any modifications you’ve already made (especially around assignment/NMS/postprocessing)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
> And i was also reading an article recently that explain the multiple label objection detection support was introduce to yolo architecture since Yolov3, you can find the article Here .

I don't know if that's true. It's simply because YOLO label format doesn't support multiple labels.

> Is it sufficient to modify the TaskAlignedAssigner to use multiple label per bounding box, or is it more complicated than that ?

You need to update TAL target scores assignment and also dataloader and data format and to support multi label.

**Comment by IGlace** [comment]
> > And i was also reading an article recently that explain the multiple label objection detection support was introduce to yolo architecture since Yolov3, you can find the article Here .
> 
> I don't know if that's true. It's simply because YOLO label format doesn't support multiple labels.
> 
> > Is it sufficient to modify the TaskAlignedAssigner to use multiple label per bounding box, or is it more complicated than that ?
> 
> You need to update TAL target scores assignment and also dataloader and data format and to support multi label.

Will just modifying the TAL, and dataloader and data format be enough ? to make the model predict multiple labels per bounding boxes.

And is this can affect the model accuracy somehow ?

**Comment by Y-T-G** [comment]
You also need to enable multi label NMS in DetectioPredictor.

**Comment by IGlace** [comment]
> You also need to enable multi label NMS in DetectioPredictor.

Do you have any idea how this can affect the model training ? or do you have any recommendations how data should be balanced in case of multiple labels per bounding box ?

**Comment by glenn-jocher** [comment]
Yes, it will change training dynamics: once you make `TaskAlignedAssigner.get_targets()` produce multi-hot `target_scores` (instead of the current one-hot scatter) and keep using BCE in `v8DetectionLoss`, each matched anchor will backprop multiple positive classes, which usually improves recall for co-occurring labels but can reduce precision (and AP) if your labels aren’t perfectly consistent or if some labels are very rare; there’s no “free” accuracy gain here and you’ll want to validate against a metric definition that matches your multi-label intent (see the current one-hot target construction in the [TaskAlignedAssigner reference](https://docs.ultralytics.com/reference/utils/tal/) and the BCE classification loss usage in the [loss reference](https://docs.ultralytics.com/reference/utils/loss/)). For balancing, the practical rule is you need enough examples for every label *and every common co-occurrence pair* on the same box; if you have heavy skew, oversample images containing rare labels and consider adding per-class weighting via `BCEWithLogitsLoss(pos_weight=...)` in the criterion (since Ultralytics currently uses plain BCE), otherwise the rare labels tend to be ignored.
